valkyrie/lib/valkyrie/storage_adapter.rb

```diff
@@ -36,6 +36,73 @@ def self.find_by(id:)
       end.find_by(id: id)
     end
 
+    class Fedora
+      attr_reader :connection
+      PROTOCOL = 'fedora://'
+      def initialize(connection:)
+        @connection = connection
+      end
+
+      # @param id [Valkyrie::ID]
+      # @return [Boolean] true if this adapter can handle this type of identifer
+      def handles?(id:)
+        id.to_s.start_with?(PROTOCOL)
+      end
+
+      # Return the file associated with the given identifier
+      # @param id [Valkyrie::ID]
+      # @return [Valkyrie::StorageAdapter::File]
+      def find_by(id:)
+        File.new(id: id, io: response(id: id))
+      end
+
+      # @param file [IO]
+      # @param model [Valkyrie::Model]
+      # @return [Valkyrie::StorageAdapter::File]
+      def upload(file:, model:)
+        # TODO: this is a very naive aproach. Change to PCDM
+        identifier = model.id.to_uri + '/original'
+        ActiveFedora::File.new(identifier) do |af|
+          af.content = file
+          af.save!
+          af.metadata.set_value(:type, af.metadata.type + [::RDF::URI('http://pcdm.org/use#OriginalFile')])
+          af.metadata.save
+        end
+        find_by(id: Valkyrie::ID.new(identifier.to_s.sub(/^.+\/\//, PROTOCOL)))
```

Comment:
Yes, it looks like we only need the URI, so that's enough.  I was mostly thinking that the Etag, etc. would be in the PUT response if we wanted that for any reason.